File: java/Strategy_Pattern/ducks/Duck.java
package java.Strategy_Pattern.ducks;

import fly.FlyBehaviour;
import quack.QuackBehaviour;

public abstract class Duck{
    QuackBehaviour quackBehaviour;
    FlyBehaviour flyBehaviour;

    public void performQuack(){
        quackBehaviour.quack();
    }

    public void performFly(){
        flyBehaviour.fly();
    }

    public void swim(){
        System.out.println("All ducks can float, even decoys    !");
    }

    public void setFlyBehaviour(FlyBehaviour fb){
        flyBehaviour = fb;
    }

    public void setQuackBehaviour(QuackBehaviour qb){
        quackBehaviour = qb;
    }

    public abstract void display();
}

File: python/strategy/ducks/duck.py
from abc import ABC, abstractmethod
from fly.fly_behaviour import FlyBehaviour

class Duck(ABC):

    def __init__(self):
        self.flyBehaviour: FlyBehaviour = None
    
    def setFlyBehaviour(self, flyBehaviour: FlyBehaviour):
        self.flyBehaviour = flyBehaviour
    
    def performFly(self):
        self.flyBehaviour.fly()

    def swim(self) -> None:
        print("I'm swimming")

    @abstractmethod
    def display(self) -> None:
        pass